### DISCUSSION TITLE
how to determine the type of a Widget and then cast it

### DISCUSSION THREAD

**Discussion by pm100** [fltk-rs/fltk-rs#1529] [General]
Converting c++ code to rust.

The c++ code looks at a widget* and does

```
    const Fl_Group *g = dynamic_cast<const Fl_Group *>(w);
    if (g) {
       ..... use the g Group pointer
```

My code now has a &Widget. 

I need to do the same but my rust brain cant work out how to do it, Ie If this Widget is a Group then give me a reference to that Group.

**Comment by MoAlyousef** [comment]
WidgetBase::from_dyn_widget and from_dyn_widget_ptr calls C++ dynamic_cast:
```rust
    if let Some(g) = fltk::group::Group::from_dyn_widget(&widget) {
        println!("{}", g.children());
    }
```

FLTK also provides a virtual function `as_group()` so you shouldn't need dynamic_cast:
```rust
    if let Some(g) = wind.as_group() {
        println!("{}", g.children());
    }
```

**Comment by pm100** [comment]
Ty. It's interesting that it's not doable in pure rust. (I spent some time trying to find a solution)

**Comment by MoAlyousef** [comment]
Rust’s polymorphism model is inherently different, so the closest thing you have is Any::downcast, but you have to pass through Any first.

**Comment by pm100** [comment]
is there a way to get the name of the class. I am converting this code https://github.com/elaforge/karya/blob/work/fltk/f_util.cc (print_children) to rust; the part that dumps out the whole widget tree. I suppose I could simply compile it as c++ code and call it via FFI but its a good exercise to convert it

**Reply by MoAlyousef** [reply]
There's a `fltk::utils::type_name(&widget)` function for debugging purposes. This prints an Itanium mangled name on posix systems. If you would like a demangled name, you could use a crate like cpp_demangle.

**Reply by pm100** [reply]
I cant find a  type_name function in util. (You mean in src/utils/mod.rs?)

**Reply by MoAlyousef** [reply]
Yes

**Reply by pm100** [reply]
so where is that function?

**Reply by MoAlyousef** [reply]
https://github.com/fltk-rs/fltk-rs/blob/56c755ebb8ad6ec39c4c3f2e6be87ce051fc8baf/fltk/src/utils/mod.rs#L145

**Reply by pm100** [reply]
OK I needed a cargo update. 

Not mangled, result is `<struct Widget_Derived<class Fl_Sys_Menu_Bar>` - perfect
